plugins/samples/set_cookie/plugin.cc

```diff
@@ -0,0 +1,42 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_set_cookie]
+#include "absl/strings/numbers.h"
+#include "proxy_wasm_intrinsics.h"
+
+class MyHttpContext : public Context {
+ public:
+  explicit MyHttpContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  FilterHeadersStatus onResponseHeaders(uint32_t headers,
+                                        bool end_of_stream) override {
+    const WasmDataPtr response_status = getResponseHeader(":status");
+    int response_code;
+    // Check if response status is part of the success range (2XX).
+    if (response_status &&
+        absl::SimpleAtoi(response_status->view(), &response_code) &&
+        (response_code / 100 == 2)) {
+      // Change the new cookie according to your needs.
+      addResponseHeader("Set-Cookie",
```

Comment:
While technically this example as it currently stands does set a cookie, I feel that it is somewhat incomplete as it does not look in the request for presence of an existing cookie, or embed / extract info from the cookie.

A more realistic example would be to use the cookie to preserve a session ID across multiple requests from the same user. So the flow would be roughly:

```
onRequestHeaders:
  check for presence of my_cookie in `Cookie` header
  if  my_cookie is present:
    extract/parse session ID value from it, and save in a field of MyHttpContext (e.g. MyHttpContext::session_id_). 
  else:
    leave MyHttpContext::session_id_ unset

onResponseHeaders:
  if MyHttpContext::session_id_ is set:
    log a message saying that current request is for existing session ID, including the session Id value in the log message
  else:
    generate new session ID
    log a message indicating that new session ID was created
    add "SetCookie" response header with new session ID
```

One thing to note from the above is that we only add `Set-Cookie` in the case where the request did not provide a cookie.

Reply:
Makes sense but I have one question...
Given that we can have many requests it would be good to storage theirs session Ids in a unique map shared between the requests and maybe to have a expiration policy to avoid an endless memory allocation for each new user.... WDYT?

Reply:
@walves-cit it should not be necessary to maintain a map of session IDs in the plugin. The way that cookies work is that the `Set-Cookie` response header instructs the client-side user agent (e.g. a browser) to store the cookie value on its side. It then echoes back the cookie value in subsequent requests. This effectively offloads storage of the cookie data (in this example, a session ID) to the client, so that the server (here, the plugin) doesn't need to maintain state.

So the steps outlined in my previous comment should be sufficient for managing cookies: if the client request doesn't contain a cookie, generate a new session ID and set it in the response using `Set-Cookie`; if the client request *does* contain a cookie, then extract the session ID from it.

WRT expiration policy of session IDs, cookie expiration times can be used for that: http://go/mdn/HTTP/Cookies#removal_defining_the_lifetime_of_a_cookie

For this example, in the case where the plugin is using `Set-Cookie` to set a new cookie, you could have it include a `Max-Age` setting to bound the cookie lifetime.

Reply:
Understood now.... I did not know that the same plugin instance lives during all the request stream.
It was not making sense to have the `MyHttpContext::session_id_` if it could be shared between the requests. But not its fine I reread the docs and understood the lifecycle of the `RootContext` and the `Contexts`.
Tks.

Reply:
Done... pls take a look